src/V3Delayed.cpp

```diff
@@ -221,448 +293,520 @@ class DelayedVisitor final : public VNVisitor {
-            return m_insertionPointp != nodep;
-        }
+        // Base name suffix for signals constructed below
+        const std::string baseName{"__" + vscp->varp()->shortName() + "__v"
+                                   + std::to_string(vscpInfo.nTmp++)};
+
+        // Unlink and capture the RHS value
+        AstNodeExpr* const capturedRhsp
+            = captureVal(scopep, nodep, nodep->rhsp()->unlinkFrBack(), "__VdlyVal" + baseName);
+
+        // Unlink and capture the LHS reference
+        AstNodeExpr* const capturedLhsp
+            = captureLhs(scopep, nodep, nodep->lhsp()->unlinkFrBack(), baseName);
+
+        // Create new flag
+        AstVarScope* const flagVscp = createTemp(flp, scopep, "__VdlySet" + baseName, 1);
+        // Set the flag at the original NBA
+        nodep->addHereThisAsNext(  //
+            new AstAssign{flp, new AstVarRef{flp, flagVscp, VAccess::WRITE},
+                          new AstConst{flp, AstConst::BitTrue{}}});
+        // Add the 'Post' scheduled commit
+        AstIf* const ifp = new AstIf{flp, new AstVarRef{flp, flagVscp, VAccess::READ}};
+        vscpInfo.flagUnique().postp->addStmtsp(ifp);
+        // Immediately clear the flag
+        ifp->addThensp(new AstAssign{flp, new AstVarRef{flp, flagVscp, VAccess::WRITE},
+                                     new AstConst{flp, AstConst::BitFalse{}}});
+        // Commit the value
+        ifp->addThensp(new AstAssign{flp, capturedLhsp, capturedRhsp});
+
+        // Delete original NBA
+        pushDeletep(nodep->unlinkFrBack());
     }
 
-    // Create a temporary variable in the scope of 'vscp',with the given 'name', and with 'dtypep'
-    // type, with the bits selected by 'sLsbp'/'sWidthp' set to 'insertp', other bits set to zero.
-    // Returns a read reference to the temporary variable.
-    AstVarRef* createWidened(FileLine* flp, AstVarScope* vscp, AstNodeDType* dtypep,
-                             AstNodeExpr* sLsbp, AstNodeExpr* sWidthp, const std::string& name,
-                             AstNodeExpr* insertp) {
-        // Create temporary variable.
-        AstVarScope* const tp = createNewVarScope(vscp, name, dtypep);
-        // Zero it
-        AstConst* const zerop = new AstConst{flp, AstConst::DTyped{}, dtypep};
-        zerop->num().setAllBits0();
-        insert(new AstAssign{flp, new AstVarRef{flp, tp, VAccess::WRITE}, zerop});
-        // Set the selected bits to 'insertp'
-        AstSel* const selp = new AstSel{flp, new AstVarRef{flp, tp, VAccess::WRITE},
-                                        sLsbp->cloneTreePure(true), sWidthp->cloneTreePure(true)};
-        insert(new AstAssign{flp, selp, insertp});
-        // This is the expression to get the value of the temporary
-        return new AstVarRef{flp, tp, VAccess::READ};
-    };
+    // Scheme::ValueQueuePartial/Scheme::ValueQueueWhole
+    template <bool Partial>
+    void prepareSchemeValueQueue(AstVarScope* vscp, VarScopeInfo& vscpInfo) {
```

Comment:
No need for Partial to be passed as a template vs just a parameter.

(I realize our styles differ in this, I'll accept if you leave it.)